Extensions/TokenAcquisition.cs

```diff
@@ -0,0 +1,188 @@
+            // Token acquisition service
+            services.Configure<AzureAdOptions>(myoptions => { });
+            services.AddTransient<ITokenAcquisition, TokenAcquisition>();
+            return services;
+        }
+    }
+
+    /// <summary>
+    /// Token acquisition service
+    /// </summary>
+    public class TokenAcquisition : ITokenAcquisition
+    {
+        /// <summary>
+        /// Constructor of the TokenAcquisition service. This requires the Azure AD Options to 
+        /// configure the confidential client application
+        /// </summary>
+        /// <param name="options">Options to configure the application</param>
+        public TokenAcquisition(IOptions<AzureAdOptions> options)
+        {
+            azureAdOptions = options.Value;
+        }
+
+        private AzureAdOptions azureAdOptions;
+
+        /// <summary>
+        /// The goal of this method is, when a user is authenticated, to add the user's account in the MSAL.NET cache
+        /// so that this token can then be used to acquire a token on On-behalf-of the user befor call to downstream APIs.
+        /// </summary>
+        /// <param name="userAccessToken">Access token used to call this Web API</param>
+        /// <example>
+        /// From configuration of the Authentication of the ASP.NET Core Web API 
+        /// <code>JwtBearerOptions option;</code>
+        /// 
+        /// Subscribe to the token validated event
+        /// <code>
+        /// options.Events = new JwtBearerEvents();
+        /// options.Events.OnTokenValidated = OnTokenValidated;
+        /// }
+        /// </code>
+        /// 
+        /// And then in the OnTokenValidated method:
+        /// <code>
+        /// private async Task OnTokenValidated(TokenValidatedContext context)
+        /// {
+        ///  JwtSecurityToken accessToken = context.SecurityToken as JwtSecurityToken;
+        ///  _tokenAcquisition.AddAccountToCache(accessToken);
+        /// }
+        /// </code>
+        /// </example>
+        public void AddAccountToCacheFromJwt(JwtSecurityToken jwtToken, IEnumerable<string> scopes)
+        {
+            string userAccessTokenForThisApi = jwtToken.RawData;
+            try
+            {
+                UserAssertion userAssertion = new UserAssertion(userAccessTokenForThisApi, "urn:ietf:params:oauth:grant-type:jwt-bearer");
+
+                // .Result to make sure that the cache is filled-in before the controller tries to get access tokens
+                AuthenticationResult result = Application.AcquireTokenOnBehalfOfAsync(scopes, userAssertion).Result;
+                string acessTokenForGraphOBOUser = result.AccessToken;
+            }
+            catch (MsalException ex)
+            {
+                string message = ex.Message;
+                throw;
+            }
+        }
+
+        public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceivedContext context, IEnumerable<string> scopes)
+        {
+            try
```

Comment:
null parameters checks?

Reply:
I'm not sure if context will ever be null, but there should be a null check for scopes at least.